### PR TITLE
Add typeclass invariant contravariant

### PR DESCRIPTION
**Pull request by ersin-ertan**
Solves: #283

### CODE REVIEW

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/InvariantLaws.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import arrow.test.generators.genFunctionAToB
+import arrow.typeclasses.Applicative
+import arrow.typeclasses.Eq
+import arrow.typeclasses.Invariant
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+
+object InvariantLaws {
+
+  inline fun <F> laws(AP: Applicative<F>, EQ: Eq<Kind<F, Int>>): List<Law> =
```
Comment (pakoito) [inline modules/core/arrow-test/src/main/kotlin/arrow/test/laws/InvariantLaws.kt]: Inline isn't necessary. It's a leftover from before a refactor.

Comment (ersin-ertan) [reply modules/core/arrow-test/src/main/kotlin/arrow/test/laws/InvariantLaws.kt]: Should I remove `inline` from the others tests too, including `noinline` for those who use it?
But leave `inline` for those tests with `reified`

**Comment by ersin-ertan** [comment]
Finished 2/3. For:
> add ContravariantLaws.test for the datatypes that implement Contravariant

Sometimes `ContravariantLaws.laws(ABC.contravariant(), ::ABC, lastParam)` the lastParam is `Eq.any()` or either `EQ()` or `EQ`. When is which required?

**Comment by pakoito** [comment]
>Sometimes ContravariantLaws.laws(ABC.contravariant(), ::ABC, lastParam) the lastParam is Eq.any() or either EQ() or EQ. When is which required?

Simple data classes like `Either`, `Option` or `ListK` have `equals` written by the kotlin compiler so `Eq.any()` is enough. Other datatypes that need to be run like Function, Eval or IO require a specialized `Eq` instance that runs them for a value.

> SortedMapKTest
ContravariantLaws.laws(SortedMapK.contravariant<Int>(), { a: Int -> sortedMapOf("key" to a).k() }, EQ)

See that the error is telling you that your map has String as the key, but the instance your passing has Int as key! `SortedMapK.contravariant<String>()` should fix it.

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/InvariantLaws.kt (hunk 1 above)
Comment (pakoito) [reply modules/core/arrow-test/src/main/kotlin/arrow/test/laws/InvariantLaws.kt]: There should be no tests with reified AFAIK. Leave the fix for a different PR tho.

**Comment by ersin-ertan** [comment]
Investigating why the tests are failing

File: modules/docs/arrow-docs/docs/docs/typeclasses/intro/README.md [hunk 2]
```diff
@@ -86,9 +86,13 @@ We will list them by their hierarchy.
 
 - [`MonoidK`]({{ '/docs/typeclasses/monoidk/' | relative_url }}) - combinable datatypes have an empty value
 
-##### Functor
+##### Invariant, Contravariant, Functor
 
-- [`Functor`]({{ '/docs/typeclasses/functor/' | relative_url }}) - its contents can be mapped
+- [`Invariant`]({{ '/docs/typeclasses/Invariant/' | relative_url }}) - its contents can be bidirectionally mapped
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/typeclasses/intro/README.md]: IIRC this is case sensitive. Please double check it!

File: modules/docs/arrow-docs/docs/docs/typeclasses/intro/README.md [hunk 3]
```diff
@@ -86,9 +86,13 @@ We will list them by their hierarchy.
...
+
+- [`Contravariant`]({{ '/docs/typeclasses/contravariant/' | relative_url }}) - its contents can be mapped reverse sequentially
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/typeclasses/intro/README.md]: To be honest I'd move these below `Bimonad`, as they're not as common as the others. Same in the menu.

Comment (ersin-ertan) [reply modules/docs/arrow-docs/docs/docs/typeclasses/intro/README.md]: The intro stated that "We will list them by their hierarchy" so I used that as the guide, however if commonality is more important, then I can switch them around, remove the quote, and change the section title back to Functor.

Comment (pakoito) [reply modules/docs/arrow-docs/docs/docs/typeclasses/intro/README.md]: Fair, up to you then :D

**Comment by ersin-ertan** [comment]
I'm a bit confused at this point. Tests are failing because `@instance` annotated interfaces with `.contramap(f: (B) -> A): Data<B> = fix().contramap(f)` and `imap` point to themselves, thus stackoverflow.

Looking at cats/data, only a few implement `contramap` and almost none implement `imap`. I think I was over-eager my imap and contramap addition, as well as defined the function parameters incorrectly. 

My next step is to remove almost all `@instance` annotated imap declarations.
My next step is to remove most all `@instance` annotated contramap declarations.
Change contramap's within the remaining `@instance` annotated contramap declarations params to include the arg that are implicit in scala.
Thoughts?

**Comment by pakoito** [comment]
You can implement `imap` in the instances in terms of `map` and `contramap` member methods, as far as I can see.

**Comment by pakoito** [comment]
A new version of Invariant and Contravariant is already in the database. Let's close this :D Thanks @ersin-ertan for the effort!

**Pull request closed without merging**